src/fancyflow.erl

```diff
@@ -1,29 +1,184 @@
 -module(fancyflow).
 
+-include_lib("syntax_tools/include/merl.hrl").
+
 %% API exports
--export([pipe/2, maybe/2, parallel/1]).
+-export([parse_transform/2]).
+-export([maybe/2, parallel/1]).
+
+-type op_kind() :: mapper | folder.
+-record(op, {kind :: op_kind()
+            ,mname :: module()
+            ,fname :: atom()
+            }).
+
+-ifdef(fun_stacktrace).
+-define(WITH_STACKTRACE(E,T,ST), E:T -> ST = erlang:get_stacktrace(),).
+-else.
+-define(WITH_STACKTRACE(E,T,ST), E:T:ST ->).
+-endif.
 
 %%====================================================================
-%% API functions
+%% Parse Transform
+%%====================================================================
+
+parse_transform(ASTs, _Options) ->
+    try [erl_syntax_lib:map(fun revert_then_transform/1, AST) || AST <- ASTs]
+    catch _E:_R -> ASTs
+    end.
+
+revert_then_transform(T) ->
+    transform(erl_syntax:revert(T)).
+
+%% [pipe](Var, F1, F2)
+%% [parallel](A,B,C,D)
+%% [{folder,knm_numbers,pipe}](T,F1,F2,F3)
+transform({call, Line,
+           {cons, CLine, Operator={_,CLine,_}, {nil,CLine}},
+           InitAndFunsOrJustFuns}) when is_list(InitAndFunsOrJustFuns) ->
+    Op = op_new(Operator),
+    op_mixin(Op, InitAndFunsOrJustFuns, Line, CLine);
+transform(Term) ->
+    Term.
+
+op_new({atom,_,pipe}) -> #op{kind=folder, mname=?MODULE, fname=pipe};
+op_new({atom,_,maybe}) -> #op{kind=folder, mname=?MODULE, fname=maybe};
+op_new({atom,_,parallel}) -> #op{kind=mapper, mname=?MODULE, fname=parallel};
+op_new({tuple, _, [{atom,_,K}, {atom,_,M}, {atom,_,F}]}) ->
+    #op{kind=K, mname=M, fname=F}.
```

Comment:
afaict this crashes hard on an unsupported form rather than just tunnelling it through